### ISSUE TITLE
Scripting the exporting step

### ISSUE THREAD

**Issue by ctanas** [kaushalmodi/ox-hugo#396]
I'm interested in scripting the exporting step for ox-hugo and I'm not sure how to do it. [I know Emacs can be called from terminal, run a function and then quit](https://www.reddit.com/r/emacs/comments/il1hue/how_can_i_open_a_file_in_emacs_run_a_macro_then/g3pishy?utm_source=share&utm_medium=web2x&context=3), but my Emacs knowledge is limited and I have no idea how to translate this to my case (I'm using a macro to call C-c C-e H A), but I have no idea how to pass this to Emacs from a script, to run it for me, without having me type that into an Emacs session.

Basically I want to write a Bash script to run C-c C-e H A for ox-hugo in Emacs, can it be done?

**Comment by vinurs** [comment]
do u need this? https://github.com/kaushalmodi/ox-hugo/issues/292
i use the Makefile, and in cli just type `make md`, it exports all post

**Comment by ctanas** [comment]
Thank you! However, it fails for me:
```
[claudiu@yhorm parsec.ro]$ make md
Emacs binary used: emacs
Loading /home/claudiu/Documents/parsec.ro/setup-ox-hugo.el (source)...
Debug on Error enabled globally
  INFO     Scraping files for ox-hugo-autoloads.el... 
  INFO     Scraping files for ox-hugo-autoloads.el...done
Loading /home/claudiu/Documents/parsec.ro/ox-hugo-autoloads.el (source)...
Debugger entered--Lisp error: (file-missing "Cannot open load file" "No such file or directory" "ox-hugo")
  require(ox-hugo)
  eval-buffer(#<buffer  *load*> nil "/home/claudiu/Documents/parsec.ro/setup-ox-hugo.el" nil t)  ; Reading at buffer position 9425
  load-with-code-conversion("/home/claudiu/Documents/parsec.ro/setup-ox-hugo.el" "/home/claudiu/Documents/parsec.ro/setup-ox-hugo.el" nil nil)
  load("/home/claudiu/Documents/parsec.ro/setup-ox-hugo.el" nil nil t)
  load-file("/home/claudiu/Documents/parsec.ro/setup-ox-hugo.el")
  (progn (setenv "OX_HUGO_TMP_DIR" "/tmp/claudiu/ox-hugo-dev") (load-file (expand-file-name "setup-ox-hugo.el" "/home/claudiu/Documents/parsec.ro")) (message "[Version check] Emacs %s" emacs-version) (message "[Version check] %s" (org-version nil :full)))
  eval((progn (setenv "OX_HUGO_TMP_DIR" "/tmp/claudiu/ox-hugo-dev") (load-file (expand-file-name "setup-ox-hugo.el" "/home/claudiu/Documents/parsec.ro")) (message "[Version check] Emacs %s" emacs-version) (message "[Version check] %s" (org-version nil :full))) t)
  command-line-1(("--eval" "(progn(setenv \"OX_HUGO_TMP_DIR\" \"/tmp/claudiu/ox-h..." "--kill"))
  command-line()
  normal-top-level()

make: *** [Makefile:141: vcheck_emacs] Error 255
```

**Comment by vinurs** [comment]
hello, i use it like this,
clone it in my vinurs.github.io directory
```
  git clone --recurse-submodules -j8 https://github.com/kaushalmodi/ox-hugo
```
then, modify the ox-hugo/Makefile
```
  diff --git a/Makefile b/Makefile
  index fc36596..0c4903e 100644
  --- a/Makefile
  +++ b/Makefile
  @@ -70,10 +70,10 @@ TEST_ENABLED=0
   OX_HUGO_TEST_DIR=$(shell pwd)/test

   # Base directory for the Hugo example site
  -OX_HUGO_TEST_SITE_DIR=$(OX_HUGO_TEST_DIR)/site
  +OX_HUGO_TEST_SITE_DIR=~/system-configuration/github/vinurs.github.io

   # Directory containing Org files for the test site
  -OX_HUGO_TEST_ORG_DIR=$(OX_HUGO_TEST_SITE_DIR)/content-org
  +OX_HUGO_TEST_ORG_DIR=~/system-configuration/self-management
```
then in ox-hugo, `make md`

**Comment by kaushalmodi** [comment]
@vinurs Thanks for helping out.

@ctanas The `make md` flow has been working fine on auto pilot in the ox-hugo Travis CI runs. Does the `make md` work fine in the git root once you clone this repo? 

From the error you posted:

> entered--Lisp error: (file-missing "Cannot open load file" "No such file or directory" "ox-hugo")

The `make vcheck` step is failing because it's not finding the ox-hugo package installation.

**Comment by vinurs** [comment]
@kaushalmodi now, the Makefile can not export org file with org-id,
```
* Placeholder Post :ATTACH:
:PROPERTIES:
:EXPORT_FILE_NAME: placeholder-post
:ID:       fe108344-8a13-41f9-80a6-f6aa86bd99a3
:END:

#+ATTR_HTML: :width 80% :align center
[[attachment:_20201020_150051screenshot.png]]
```
like this, the png file is stored in my `org-directory`/.attach/fe/108344-8a13-41f9-80a6-f6aa86bd99a3, use `make md` can not find the ATTACH
```
Debugger entered--Lisp error: (org-link-broken "attachment:_20201020_151326screenshot.png")
  signal(org-link-broken ("attachment:_20201020_151326screenshot.png"))
  org-export-resolve-fuzzy-link((link (:type "fuzzy" :path "attachment:_20201020_151326screenshot...." :format bracket :raw-link "attachment:_20201020_151326screenshot...." :application nil :search-optio
```

**Comment by kaushalmodi** [comment]
> now, the Makefile can not export org file with org-id

What changed? I haven't made an update to ox-hugo in a while. 

I have few org-id tests in the ox-hugo test suite and they all have been passing every night. 

Does that export work when manually exporting using ox-hugo and not using the Makefile?

**Comment by vinurs** [comment]
i use doom-emacs, and in doom emacs set org-directory
```
(setq org-directory "~/system-configuration/pdms/")
```
and all screenshot files in `org-directory`/.attach directory，so when `make md`, it can not find the file

**Comment by kaushalmodi** [comment]
> i use doom-emacs, and in doom emacs set org-directory ..

Does that mean that the export works fine when you use your emacs config and export the file manually (`C-c C-e H H`)?

**Comment by vinurs** [comment]
> > i use doom-emacs, and in doom emacs set org-directory ..
> 
> Does that mean that the export works fine when you use your emacs config and export the file manually (`C-c C-e H H`)?

yes

**Comment by kaushalmodi** [comment]
well.. they you just need to set the `org-directory` variable in the Elisp setup file that you might be loading in your Makefile.

For example, for the ox-hugo nightly CI and doc generation, I have this setup file (I've linked to an example of how I set org variables I need for the tests and doc generation to work as I expect): https://github.com/kaushalmodi/ox-hugo/blob/1909c6effe578cf41fcf436da881f1cd430a498a/test/setup-ox-hugo.el#L214

**Comment by vinurs** [comment]
> generation

i tried it just now, but the same error

**Comment by vinurs** [comment]
> well.. then you just need to set the `org-directory` variable in the Elisp setup file that you might be loading in your Makefile.
> 
> For example, for the ox-hugo nightly CI and doc generation, I have this setup file (I've linked to an example of how I set org variables I need for the tests and doc generation to work as I expect):
> 
> https://github.com/kaushalmodi/ox-hugo/blob/1909c6effe578cf41fcf436da881f1cd430a498a/test/setup-ox-hugo.el#L214

i solve this, it should add this in `setup-ox-hugo.el`
```
(require 'org-attach)
```
thanks for you help

**Comment by kaushalmodi** [comment]
@ctanas Let me know if you need any help with this issue, or if it can be closed?

**Comment by vinurs** [comment]
@kaushalmodi another problem, run `make md`，but it doesn't copy the file to `static/ox-hugo` directory

**Comment by ctanas** [comment]
> The `make vcheck` step is failing because it's not finding the ox-hugo package installation.

This is weird, since it exports fine from inside Emacs. ox-hugo package is clearly installed, maybe it's just not found by make?

**Comment by kaushalmodi** [comment]
> maybe it's just not found by make?

That's correct. Check if the variables for paths to install and look for the ox-hugo package are set up correctly in the Makefile.

**Comment by kaushalmodi** [comment]
@vinurs 

> another problem, run make md，but it doesn't copy the file to static/ox-hugo directory

Again, if this problem happens with `make md` but not with the interactive `C-c C-e H H`, please review what Emacs/Org variable settings are missing/wrong in the Makefile loaded Elisp setup file.

You simply need to set the necessary elisp variables in the Makefile loaded setup file that you would regularly rely on in your interactive Emacs sessions.

See all the setup I do in the [`setup-ox-hugo.el`][1] that I use for the nightly Travis CI and ox-hugo doc generation.

[1]: https://github.com/kaushalmodi/ox-hugo/blob/master/test/setup-ox-hugo.el

**Comment by vinurs** [comment]
@kaushalmodi thanks, it should add the following in `setup-ox-hugo.el`
```
+  (setq org-directory "~/system-configuration/pdms/")
+ (setq org-attach-id-dir (expand-file-name ".attach/" org-directory))
```
it works ok now.

**Comment by bphenriques** [comment]
I have this working on my personal knowledge-base. It is as minimal as I can:
https://github.com/bphenriques/knowledge-base/blob/master/tools/init.el

https://github.com/bphenriques/knowledge-base/blob/master/Makefile#L18

Hope it helps

**Comment by kaushalmodi** [comment]
I am closing this issue as it appears to be resolved.
